common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -5718,6 +5718,72 @@ static int cert_verify_callback(X509_STORE_CTX* x509_store_ctx, void* arg) {
     return result;
 }
 
+static int cert_verify_callback2(X509_STORE_CTX *x509_store_ctx, void *arg) {
+    /* Get the correct index to the SSLobject stored into X509_STORE_CTX. */
+    SSL* ssl = reinterpret_cast<SSL*>(X509_STORE_CTX_get_ex_data(x509_store_ctx,
+            SSL_get_ex_data_X509_STORE_CTX_idx()));
+    JNI_TRACE("ssl=%p cert_verify_callback2 x509_store_ctx=%p arg=%p", ssl, x509_store_ctx, arg);
+
+    AppData* appData = toAppData(ssl);
+    JNIEnv* env = appData->env;
+    if (env == nullptr) {
+        ALOGE("AppData->env missing in cert_verify_callback2");
+        JNI_TRACE("ssl=%p cert_verify_callback2 => 0", ssl);
+        return 0;
+    }
+    // Get a stack of all certs in the chain
+    STACK_OF(X509) *sk = x509_store_ctx->untrusted;
+
+    int len = sk_X509_num(sk);
+
+    // Create the byte[][] array that holds all the certs
+    ScopedLocalRef<jobjectArray> array(env,
+        env->NewObjectArray(len, JniConstants::byteArrayClass, nullptr));
+
+    for(unsigned i = 0; i < len; ++i) {
+        X509* cert = sk_X509_value(sk, i);
+
+        unsigned char* buf = nullptr;
+        int length = i2d_X509(cert, &buf);
+        if (length < 0) {
```

Comment:
Just call `SSL_get0_peer_certificates`. Those will be pre-encoded and get you marginally closer to the happy `X509`-less future world. (OpenSSL's X.509 library is incredibly inefficient. We saw speed, memory, and binary size improvements when we switched Chromium away from it. There's a lot of work to be done before Conscrypt can realize the same wins, but one step at a time.)